src/app/data-model-provider/ActionReturnStatus.h

```diff
@@ -0,0 +1,92 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/support/StringBuilder.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <variant>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/// An ActionReturnStatus encodes the result of a read/write/invoke.
+///
+/// Generally such actions result in a StatusIB in the interaction model,
+/// which is a code (InteractionModel::Status) and may have an associated
+/// cluster-specific code.
+///
+/// However some actions specifically may return additional errors for
+/// chunking (i.e. out of space, retry or send existing partial data) hence
+/// the exitance of this class:
+///
+///   - encapsulates a ClusterStatusCode for an actual action result
+///   - encapsulates a underlying CHIP_ERROR for reporting purposes
+///     and to check for out-of-space specific errors
+///
+/// The class is directly constructible from statuses (non-exlicit) to make
+/// returning of values easy.
+class ActionReturnStatus
+{
+public:
+    ActionReturnStatus(CHIP_ERROR error) : mReturnStatus(error) {}
+    ActionReturnStatus(Protocols::InteractionModel::Status status) :
+        mReturnStatus(Protocols::InteractionModel::ClusterStatusCode(status))
+    {}
+    ActionReturnStatus(Protocols::InteractionModel::ClusterStatusCode status) : mReturnStatus(status) {}
+
+    /// Constructs a status code. Either returns the underlying code directly
+    /// or converts the underlying CHIP_ERROR into a cluster status code.
+    Protocols::InteractionModel::ClusterStatusCode GetStatusCode() const;
+
+    /// Gets the underlying CHIP_ERROR if it exists, otherwise it will
+    /// return a CHIP_ERROR corresponding to the underlying return status.
+    ///
+    /// Success statusess will result in CHIP_NO_ERROR (i.e. cluster specitic success codes are lost)
+    CHIP_ERROR GetUnderlyingError() const;
+
+    /// If this is a CHIP_NO_ERROR or a Status::Success
+    bool IsSuccess() const;
+
+    /// Considers if the underlying error is an error or not (CHIP_NO_ERROR is the only non-erro)
+    /// or if the underlying statuscode is not an error (success and cluster specific successes
+    /// are not an error).
+    bool IsError() const { return !IsSuccess(); }
+
+    /// Checks if the underlying error is an out of space condition (i.e. something that
+    /// chunking can handle by sending partial list data).
+    ///
+    /// Generally this is when the return is based on CHIP_ERROR_NO_MEMORY or CHIP_ERROR_BUFFER_TOO_SMALL
```

Comment:
I don't think we should document these specific status codes, because I still think we should move toward not representing "needs chunking" via magic error codes at any point other than the immediate TLVWriter API boundary, in followups to this PR.

I would consider naming this OutOfSpaceToEncodeResponse(), by the way.

Reply:
Renamed.